src/SignalR/server/SignalR/test/WebSocketsTransportTests.cs

```diff
@@ -49,6 +58,33 @@ public void HttpOptionsSetOntoWebSocketOptions()
             Assert.Same(httpOptions.Credentials, webSocketsOptions.Credentials);
         }
 
+        [ConditionalFact]
+        [WebSocketsSupportedCondition]
+        public async Task HttpOptionsWebSocketFactoryIsUsed()
+        {
+            var httpOptions = new HttpConnectionOptions();
+            var webSocketMock = new Mock<WebSocket>();
+            bool factoryWasUsed = false;
+
+            // we emulate that connection is closed right away after it was established
+            webSocketMock.Setup(socket => socket.CloseStatus).Returns(WebSocketCloseStatus.NormalClosure);
+            webSocketMock.Setup(socket => socket.ReceiveAsync(It.IsAny<Memory<byte>>(), It.IsAny<CancellationToken>()))
+                .ReturnsAsync(new ValueWebSocketReceiveResult(0, WebSocketMessageType.Close, true));
+
+            httpOptions.WebSocketFactory = (context, token) =>
```

Comment:
Could you capture the options here and do a comparison at the end of the test with the original `HttpConnectionOptions`. `Assert.NotSame(httpOptions, capturedOptions);` and some property comparisons, like add some headers and make sure they exist.

Reply:
also want to confirm. so shallow copying happens on `HttpConnectionFactory` level and if we are testing against `WebSocketsTransport`, then it won't do shallow copying. i can add tests to verify that `options` object that we used to construct `WebSocketsTransport` is the same that we got in the `WebSocketFactory` but it would be specific use case. 

And if we would have full end to end test that would create HubConnection\HttpConnection, then it would do shallow copy. 
as another possible way i can also add another test to `Microsoft.AspNetCore.SignalR.Client.Tests` to HttpConnectionTests.Transport.cs or some other HubConnectionTests so it would test shallow copy and verify that user object wasn't changed